### PR TITLE
Add copy option -u to edit-chroot

### PR DESCRIPTION
**Pull request by justinguy**
Adds the option to copy a chroot using `edit-chroot -u DEST`. See #2168

### CODE REVIEW

**Comment by justinguy** [comment]
Just signed the Individual Contributor License Agreement per the contributor section of `README.md`.

File: host-bin/edit-chroot [hunk 1]
```diff
@@ -57,6 +58,12 @@ Options:
                 If multiple chroots are specified, DEST must be a directory.
                 If you are moving a chroot to a SD card/USB drive, make sure the
                 storage is formatted to ext2/3/4.
+    -u DEST     Copies a chroot. Specify a new name to keep it in the same
```
Comment (dnschneid) [inline host-bin/edit-chroot]: `Like -m, but copies instead of moves.`

File: host-bin/edit-chroot [hunk 2]
```diff
@@ -153,11 +166,13 @@ if [ -n "$BACKUP$RESTORE" -a -z "$TARBALL" -a "$PWD" = '/' \
...
 fi
 
-# If multiple chroots are listed, KEYFILE and MOVE must be empty or directories.
+# If multiple chroots are listed, KEYFILE and MOVE and COPY must be empty or directories.
 if [ $# -gt 1 -a -f "$KEYFILE" -a ! "$KEYFILE" = '-' ]; then
     error 2 "Multiple chroots specified, but $KEYFILE is not a directory."
 elif [ $# -gt 1 -a -n "$MOVE" -a "${MOVE%/}" = "$MOVE" ]; then
     error 2 "Multiple chroots specified, but $MOVE is not a directory."
+elif [ $# -gt 1 -a -n "$COPY" -a "${COPY%/}" = "$COPY" ]; then
+      error 2 "Multiple chroots specified, but $COPY is not a directory."
```
Comment (dnschneid) [inline host-bin/edit-chroot]: indentation

File: host-bin/edit-chroot [hunk 3]
```diff
@@ -594,6 +615,31 @@ for NAME in "$@"; do
...
+        fi
+        if [ -e "$target" ]; then
+            # Can't tell if the destination is a directory or a chroot that
+            # already exists; be safe and assume it was a mistake.
+            error 2 "$target already exists"
+        fi
+        if ! mkdir -p "`dirname "$target"`"; then
+            error 1 "Unable to create directory for $target"
+        fi
+        echo "Copying $CHROOT to $target" 1>&2
```
Comment (dnschneid) [inline host-bin/edit-chroot]: Use the spinner approach that the cp portion of MOVE uses

**Comment by justinguy** [comment]
Thanks for the feedback @dnschneid working on Sunday I see! Just pushed my updates. I also added the `--one-file-system` option because you had it in `-m`. I understand what the option does but curious as to what the reason was to use it. Cheers!

**Comment by dnschneid** [comment]
Looking good.  `--one-file-system` is to protect the script from any weird bind mounts that people may be doing.  Don't leave home without it :)

Could you add tests to `test/tests/12-edit-chroot` to exercise the new feature?  Make sure to poke at the edge cases and error conditions.

**Comment by justinguy** [comment]
I will work on the tests in the next couple days.

**Comment by justinguy** [comment]
Added the tests. Traveling with my just my Chromebook so resorted to running the tests manually, let me know if any problems. Thanks!

**Comment by dnschneid** [comment]
Oops, sorry, I just remembered that we want to add -u for updating chroots (#2115).  How about using -C (capital C)?

**Comment by justinguy** [comment]
No problem, should have seen that  for -u. Changed to `-C` and added a test for trying to copy and move together.

**Comment by justinguy** [comment]
Is there pointers somewhere for setting up a proper development environment so I can run the tests easier in the future? I usually don't work this far down the stack. Thanks!

**Comment by dnschneid** [comment]
Thanks!  The tests are designed to be run straight from crosh on a Chromebook (not inside a chroot).  Copy the state of your repo to the chromebook (or use wget to grab the .tar.gz archive of your github repo), then use test/run.sh with some prefix of tests you want to run (such as 12 for 12-edit-chroot).

Other than that, it's up to you whether you want to develop in a chroot on the same machine you're testing on.

**Comment by justinguy** [comment]
I figured that was the case. I am traveling and only carrying my Pixel. Combined with Crouton it is absolutely amazing. However I am reluctant to run tests (with things like filling the disk) because restoring everything from a powerwash takes forever. I guess if I want to contribute more I could just pick up a cheap Chromebook for testing.

Thanks for everything @dnschneid

**Comment by dnschneid** [comment]
If you limit to just running one test (12 in this case), it won't run out of disk space, and the runner does a good job of cleaning up after itself on error.  Anyway, I ran the test:

```
515 [t] Expecting 'host edit-chroot -y -C /usr/local/test-20151028081933 prefix/12-edit-chroot.precise.1.Cwv/chroots.2 precise precise.2' to fail
515 [t] LAUNCHING: edit-chroot -y -C /usr/local/test-20151028081933 prefix/12-edit-chroot.precise.1.Cwv/chroots.2 precise precise.2
515 [e] Multiple chroots specified, but /usr/local/test-20151028081933 prefix/12-edit-chroot.precise.1.Cwv/chroots.2 is not a directory.
515 [t] FAIL with code 2: edit-chroot -y -C /usr/local/test-20151028081933 prefix/12-edit-chroot.precise.1.Cwv/chroots.2 precise precise.2
515 [t] PASS: 'host edit-chroot -y -C /usr/local/test-20151028081933 prefix/12-edit-chroot.precise.1.Cwv/chroots.2 precise precise.2' returned 2
515 [t] LAUNCHING: enter-chroot -n precise true
515 [e] /usr/local/test-20151028081933 prefix/12-edit-chroot.precise.1.Cwv/chroots/precise not found.
515 [t] FAIL with code 1: enter-chroot -n precise true
```

**Comment by justinguy** [comment]
Went through the testing suite and got more comfortable with running them. Thanks for your patience. Just committed tests that are passing. This ended up being a deeper exercise than I expected, but it is awesome and I have learned a lot!

File: host-bin/edit-chroot [hunk 4]
```diff
@@ -594,6 +610,31 @@ for NAME in "$@"; do
             mv "$CHROOT" "$target"
         fi
     fi
+
+    # Copy the chroot if requested
```
Comment (dnschneid) [inline host-bin/edit-chroot]: Actually, shouldn't this go above the MOVE section?  That way you can copy and move at the same time...

Comment (justinguy) [inline host-bin/edit-chroot]: I actually built it so you can't copy and move at the same time, and included testing for this. I can't see the use case of wanting to do both, and it seems it would be more likely to be a mistake. Do you feel this needs to be reworked so that you can move while copying?

**Comment by dnschneid** [comment]
Finally merged.  Thanks!
